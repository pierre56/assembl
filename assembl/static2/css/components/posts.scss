.translate {
  font-size: $font-size-xs;
  font-family: $font-family6;
  display: flex;
  align-items: center;

  .translate-button {
    margin: 5px;
  }
}

.choose-content-locale-form {
  .modal-body {
    text-align: left !important;
  }
}

.proposals .post-body .body,
.posts .post-body .body {
  font-family: $font-family4;
  color: $post-text-color;
  line-height: 1.6em;
  word-wrap: break-word;
  text-align: left;
}

.proposals {
  .custom-overlay {
    display: inline-block;
  }

  .box {
    border-bottom: $navbar-border;
    border-left: $border-theme;
    border-right: $navbar-border;
    border-top: $navbar-border;
    margin: 0 auto 40px;
    max-width: 900px;
    padding: 0;
    position: relative;
    text-align: center;
  }

  .statistics-container {
    height: 70%;
  }

  .user {
    text-align: left;

    .username {
      color: $first-color;
    }
  }

  .translate {
    padding: 20px 0 0 75px;
  }

  .post-footer {
    border-top: $navbar-border;
    padding: 15px 0 0;
    display: flex;
    justify-content: space-between;

    .sentiments {
      font-family: $font-family4;
    }

    .actions {
      margin: 5px;
    }
  }

  .overflow-action {
    cursor: pointer;
    margin: 20px 0 0;
    color: $first-color-opacity-70;

    &:hover {
      color: $first-color;
    }
  }

  .stat-sentiment {
    padding: 0 0 0 3px;
  }

  .min-sentiment {
    display: inline-block;
    font-family: $font-family4;
    font-size: $font-size-xxs;
    text-align: left;
    width: 60px;

    svg {
      vertical-align: bottom;
    }

    .txt {
      font-size: 12px;
      line-height: 12px;
      padding: 0 0 0 5px;
    }
  }

  .totalSentimentsCount {
    font-family: $font-family3;
    font-size: $font-size-sm;
    letter-spacing: initial;
    margin: 0 0 0;
    position: absolute;
    top: 30%;
    width: 100%;
  }

  .sentimentsCountLabel {
    font-family: $font-family4;
    font-size: $font-size-xxs;
    letter-spacing: initial;
    margin: 20px 0 0;
    position: absolute;
    top: 70px;
    width: 100%;
  }

  .question-footer {
    margin: 0 0 70px;
    display: flex;
    justify-content: center;

    :first-child {
      margin-left: 0;
    }

    span {
      margin-left: 5px;
    }
  }
}

$deleted-post-border: 1px solid $grey1 !default;

.idea .level .deleted-post .box {
  border-left: $deleted-post-border;
}

.idea .level .box.pending {
  border-left: 4px solid #ffd48f;
}

.posts {
  &.deleted-post {
    .body {
      font-style: italic;
      font-family: $font-family;
    }
  }

  .embedded-video {
    width: 100%;
    height: 300px;
  }

  @media screen and (max-width: 500px) {
    .embedded-video {
      height: 250px;
    }
  }

  @media screen and (max-width: 321px) {
    .embedded-video {
      height: 180px;
    }
  }

  .post-row {
    display: flex;
  }

  .post-left {
    width: 93%;
    padding: 40px;
  }

  .post-right {
    width: 7%;
    position: relative;
    text-align: center;
    background-color: $first-color-light;
    min-width: 40px;

    .answers {
      display: none;
    }

    .post-actions {
      width: 100%;
    }
  }

  .post-right.pending {
    background-color: #ffd48f;
  }

  .translate {
    padding: 20px 0 0;
  }

  .box {
    border-bottom: $navbar-border;
    border-left: $border-theme;
    border-right: $navbar-border;
    border-top: $navbar-border;
    margin: 0 auto;
    max-width: 600px;
    position: relative;
    text-align: left;
    padding: 0;

    .label {
      display: block;
      white-space: normal;
      text-align: left;
      line-height: 12px;
    }
  }

  .profileLine {
    img {
      vertical-align: text-bottom;
    }

    .creator {
      color: $second-color;
      padding: 0 0 0 10px;
    }

    .creator.pending {
      color: #ffd48f;
      text-transform: uppercase;
    }

    .date {
      color: $grey2;
      font-size: $font-size-xs;
      padding: 0 0 0 10px;
      font-family: $font-family4;
    }
  }

  .body {
    padding: 20px 0 40px;

    p {
      margin: 0;
      min-height: 1em; /* so that empty paragraphs (created by multiple line returns) still have height */
    }
  }

  .link-idea {
    border-bottom: $navbar-border;
    border-top: $navbar-border;
    margin: 0 0 5px 0;
    padding: 5px 0;

    .label {
      color: $grey2;
      font-size: $font-size-xs;
      font-family: $font-family4;
    }

    .badges {
      padding: 5px 0;

      .badge {
        display: inline;
        background-color: $grey7;
        color: $grey3;
        text-align: left;
        white-space: normal;
        word-wrap: break-word;
        font-family: $font-family4;
      }
    }
  }

  .post-action {
    cursor: pointer;
    margin: 10px 0 0;
    transition: all 0.2s ease-in-out;
    border: none;
    background-color: inherit;

    &:hover {
      transform: scale(1.2);
    }
  }

  .overflow-action {
    cursor: pointer;
    margin: 20px 0 0;
    color: $first-color-opacity-70;

    &:hover {
      color: $first-color;
    }

    .assembl-icon-ellipsis-vert {
      &:before {
        margin-left: 0;
        margin-right: 0;
        width: 100%;
        padding-left: 3px;
      }
    }
  }

  .add-sentiment {
    margin: 20px 0 0;

    .sentiment {
      margin-left: auto;
      margin-right: auto;
    }
  }

  .answer-form {
    background-color: $first-color-light;
    border-top: 2px solid $white;
    padding: 40px;

    .assembl-icon-back-arrow {
      &:before {
        font-size: $font-size-s;
      }
    }

    .answer-form-inner {
      margin: 13px;

      .editor-label {
        padding-top: 16px;
      }
    }

    .button-container {
      margin: 25px 0 0;
    }

    .assembl-icon-back-arrow {
      padding-left: 10px;
    }

    .btn {
      margin: 3px 0 8px;
    }
  }

  .collapsed-answer-form {
    background-color: $first-color-light;
    border-top: 2px solid $white;
    padding: 0 20px;

    .editor-header {
      display: none;
    }

    .public-DraftEditor-content {
      height: 20px;
    }

    .button-container {
      display: none;
    }

    .annotation {
      display: none;
    }
  }

  .extracts {
    position: absolute;
    left: 2vw;
    max-width: 230px;
    width: 14vw;
    line-height: 30px;

    .badges {
      line-height: 26px;
      font-family: $font-family4;

      .nugget-img {
        text-align: center;
        margin-bottom: 25px;

        span {
          display: inline-block;
        }

        span::before {
          font-size: 50px;
        }
      }

      .nugget {
        .nugget-txt {
          font-family: $font-family6;
          color: $first-color;
          margin-bottom: 25px;
        }

        .box-hyphen {
          border-top-width: 3px;
        }
      }
    }
  }
}

@media screen and (min-width: 600px) {
  .proposals {
    .sentiment-label {
      display: inline-block;
    }

    .stat-sentiment {
      margin: -15px 0 0;
    }

    .content {
      float: left;
      text-align: left;
      width: 80%;

      .user {
        padding: 30px 0 0 40px;
      }
    }

    .statistic {
      background-color: $first-color-light;
      float: left;
      height: 100%;
      min-width: 150px;
      padding: 0 0 0;
      position: absolute;
      right: 0;
      width: 20%;

      .statistics-doughnut {
        height: 100px;
        margin-top: 30px;
        margin-bottom: 30px;

        .circle {
          stroke-width: 14px;
        }
      }
    }

    .body {
      padding: 20px 50px 0 73px;
    }

    .post-footer {
      margin: 25px 50px 10px 73px;
    }

    .box {
      min-height: 230px;
    }
  }
}

@media screen and (max-width: 599px) {
  .proposals {
    .content {
      .post-footer {
        margin: 25px 20px 10px 25px;

        div {
          display: none;
        }
      }
    }

    .statistic {
      text-align: right;
      margin-bottom: 10px;
      display: flex;
      justify-content: space-between;

      .sentiments {
        border-top: 0;
        padding: 0;
        padding-top: 12px;
        margin-left: 10px;

        .sentiment {
          margin-left: 25px;
          vertical-align: middle;

          svg {
            vertical-align: middle;
          }
        }
      }

      .statistics-doughnut {
        height: 50px;
        width: 50px;
        margin-right: 10px;
        vertical-align: middle;

        .circle {
          stroke-width: 20px;
        }

        .doughnut-label {
          display: none;
        }
      }

      .stat-sentiment {
        margin: 0;
        margin-right: 10px;
        padding: 0;
        vertical-align: middle;
        display: inline-block;
      }

      .actions {
        margin: 0;
        padding: 0;
        margin-right: 10px;
        height: 50px;
        display: flex;
        flex-direction: column;
        justify-content: center;
      }
    }

    .content {
      text-align: left;

      .user {
        padding: 30px 20px 0;
      }
    }

    .translate {
      padding: 20px 0 0 54px;
    }

    .statistics-container {
      display: none;
    }

    .sentiment-label {
      display: none;
    }

    .totalSentimentsCount {
      display: none;
    }

    .sentimentsCountLabel {
      display: none;
    }

    .body {
      padding: 20px 20px 0 25px;
    }

    .min-sentiment {
      margin: 0 0 5px 0;
    }

    .clear {
      height: 10px;
    }
  }
}

@media screen and (min-width: 993px) {
  .posts {
    .post-actions {
      position: absolute;
      bottom: 0;
    }

    .post-right {
      min-height: 320px;
    }
  }
}

@media screen and (max-width: 992px) {
  .posts {
    .custom-overlay {
      display: inline-block;
    }

    .post-row {
      display: block;
    }

    .post-right {
      background: none;
      width: 100%;

      .answers {
        float: right;
        display: block;
        padding: 20px 20px 0 0;
      }
    }

    .post-left {
      width: 100%;

      .answers {
        display: none;
      }
    }

    .post-actions {
      position: relative;

      .post-icons {
        border-bottom: 1px solid $grey1;
        border-top: 1px solid $grey1;
        padding: 18px 0 10px;

        .post-action {
          display: inline-block;
          margin: 0 10px;
        }

        .add-sentiment {
          display: inline-block;
          margin: 0;

          .sentiment {
            margin: 0 10px 5px 10px;
          }
        }
      }
    }

    .link-idea {
      border-bottom: none;
      margin: 0;
      padding: 20px 0 0;
    }

    .sentiments-count {
      float: left;

      .txt {
        margin-left: 20px;
        text-align: left;
        padding: 1px 0 0 20px;
      }
    }

    .min-sentiment {
      padding: 1px 0 0 10px;
    }

    .answer-form {
      padding: 25px 30px 10px 30px;

      .button-container {
        .btn {
          float: none;
          width: 100%;
        }
      }

      .answer-form-inner {
        margin: 13px 0 0;
      }

      .assembl-icon-back-arrow {
        padding: 0;
      }
    }

    .extracts {
      display: none;
    }
  }
}

.column-post {
  .post-body-title {
    display: none;
  }

  .box {
    border: none;
  }

  .answer-form {
    border-top: none;
  }

  .column-hint {
    font-family: $font-family4;
    font-size: $font-size-m;
    letter-spacing: 1px;
    line-height: 30px;
    padding-top: 20px;
  }
}

.action-validate {
  border: none;
}
