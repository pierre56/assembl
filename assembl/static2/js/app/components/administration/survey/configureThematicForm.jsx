// @flow
import React from 'react';
import { I18n } from 'react-redux-i18n';
import { Field } from 'react-final-form';
import { connect } from 'react-redux';

import { get, goTo } from '../../../utils/routeMap';
import MultilingualTextFieldAdapter from '../../form/multilingualTextFieldAdapter';
import MultilingualRichTextFieldAdapter from '../../form/multilingualRichTextFieldAdapter';
import FileUploaderFieldAdapter from '../../form/fileUploaderFieldAdapter';
import SelectFieldAdapter from '../../form/selectFieldAdapter';
import { deleteThematicImageTooltip } from '../../common/tooltips';
import Helper from '../../common/helper';
import type { ThemesAdminValues, ThemeValue, ThemesValue } from './types.flow';
import { PHASES, MESSAGE_VIEW, modulesTranslationKeys } from '../../../constants';
import SurveyFields from './surveyFields';

type Props = {
  editLocale: string,
  thematicId: string,
  slug: string,
  values: ?ThemesAdminValues
};

export function getFieldData(themeId: string, values: ThemesValue, fieldName: string): { name: string, value: ?ThemeValue } {
  let result = { name: '', value: null };
  let index = 0;
  let value = values[index];
  while (!result.name && value) {
    if (value.id === themeId) {
      result = { name: `${fieldName}[${index}]`, value: value };
    } else if (value.children) {
      const childrenResult = getFieldData(themeId, value.children, 'children');
      if (childrenResult.name) {
        result = { name: `${fieldName}[${index}].${childrenResult.name}`, value: value[childrenResult.name] };
      }
    }
    index += 1;
    value = values[index];
  }
  return result;
}

class ConfigureThematicForm extends React.PureComponent<Props> {
  getName = () => {
    const { values, thematicId, slug } = this.props;
    const fieldData = getFieldData(thematicId, values ? values.themes : [], 'themes');
    if (!fieldData.name) {
      goTo(get('administration', { slug: slug, id: PHASES.survey }, { section: 1 }));
    }
    return fieldData;
  };

  render() {
    const { editLocale } = this.props;
    const { name, value: theme } = this.getName();
    const upperCaseLocale = editLocale.toUpperCase();
    const titleName = `${name}.title`;
    const descriptionName = `${name}.description`;
    const imageName = `${name}.img`;
    const messageViewOverrideName = `${name}.messageViewOverride`;
    const announcementTitleName = `${name}.announcement.title`;
    const announcementBodyName = `${name}.announcement.body`;
    return (
      <div className="form-container" key={name}>
        <Helper
          label={I18n.t('administration.headerTitle')}
          helperUrl="/static2/img/helpers/helper1.png"
          helperText={I18n.t('administration.tableOfThematics.bannerHeader')}
          classname="title"
        />
        <Field
          required
          editLocale={editLocale}
          name={titleName}
          component={MultilingualTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.thematicTitle')} ${upperCaseLocale}`}
        />
        <Field
          key={`${descriptionName}-${editLocale}`}
          editLocale={editLocale}
          name={descriptionName}
          component={MultilingualTextFieldAdapter}
          label={`${I18n.t('administration.tableOfThematics.bannerSubtitleLabel')} ${upperCaseLocale}`}
        />
        <Field
          deleteTooltip={deleteThematicImageTooltip}
          name={imageName}
          component={FileUploaderFieldAdapter}
          label={I18n.t('administration.tableOfThematics.bannerImagePickerLabel')}
        />
        <div className="title">{I18n.t('administration.tableOfThematics.moduleTypeLabel')}</div>
        <Field
          name={messageViewOverrideName}
          component={SelectFieldAdapter}
          isSearchable={false}
          // label={I18n.t('administration.tableOfThematics.moduleTypeLabel')}
          options={modulesTranslationKeys.map(key => ({ value: key, label: I18n.t(`administration.modules.${key}`) }))}
        />
        <Helper
          label={I18n.t('administration.instructions')}
          helperUrl="/static2/img/helpers/helper_BM_1.png"
          helperText={I18n.t('administration.tableOfThematics.instructionHeader')}
          classname="title"
        />
        <Field
          key={`${announcementTitleName}-${editLocale}`}
          editLocale={editLocale}
          name={announcementTitleName}
          label={`${I18n.t('administration.tableOfThematics.sectionTitleLabel')} ${upperCaseLocale}`}
          component={MultilingualTextFieldAdapter}
          required
        />
        <Field
          key={`${announcementBodyName}-${editLocale}`}
          editLocale={editLocale}
          name={announcementBodyName}
          label={`${I18n.t('administration.tableOfThematics.instructionLabel')} ${upperCaseLocale}`}
          withAttachment
          component={MultilingualRichTextFieldAdapter}
        />
        {theme && theme.messageViewOverride.value === MESSAGE_VIEW.survey ? (
          <SurveyFields editLocale={editLocale} fieldPrefix={name} />
        ) : null}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  slug: state.debate.debateData.slug
});

export default connect(mapStateToProps)(ConfigureThematicForm);