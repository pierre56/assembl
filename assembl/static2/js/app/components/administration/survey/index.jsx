// @flow
import React from 'react';
import isEqualWith from 'lodash/isEqualWith';
import { type ApolloClient, compose, withApollo } from 'react-apollo';
import { connect } from 'react-redux';
import arrayMutators from 'final-form-arrays';

import AdminForm from '../../form/adminForm';
import { compareEditorState } from '../../form/utils';
import LoadSaveReinitializeForm from '../../form/LoadSaveReinitializeForm';
import Navbar from '../navbar';
import Step1 from './step1';
import Export from './export';
import ConfigureThematicForm from './configureThematicForm';
import { load, postLoadFormat } from './load';
import { createMutationsPromises, save } from './save';
import validate from './validate';
import Loader from '../../common/loader';

type Props = {
  client: ApolloClient,
  phaseIdentifier: string,
  section: string,
  thematicId: string,
  debateId: string,
  discussionPhaseId: string,
  editLocale: string,
  locale: string
};

const loading = <Loader />;

const steps = ['1', '2'];

const DumbSurveyAdminForm = ({
  client,
  phaseIdentifier,
  section,
  thematicId,
  discussionPhaseId,
  debateId,
  editLocale,
  locale
}: Props) => {
  if (!discussionPhaseId) {
    return loading;
  }
  return (
    <LoadSaveReinitializeForm
      load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy, discussionPhaseId, locale)}
      loading={loading}
      postLoadFormat={postLoadFormat}
      createMutationsPromises={createMutationsPromises(client, discussionPhaseId)}
      save={save}
      validate={validate}
      mutators={{
        ...arrayMutators
      }}
      render={({ handleSubmit, submitting, values, initialValues }) => {
        // the form is splitted to multiple subforms: We must calculate the global pristine.
        // Don't use final form pristine here
        const pristine = isEqualWith(initialValues, values, compareEditorState);
        return (
          <React.Fragment>
            <div className="admin-content">
              <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting}>
                {section === '1' && <Step1 editLocale={editLocale} locale={locale} discussionPhaseId={discussionPhaseId} />}
                {section === 'configThematics' && (
                  <ConfigureThematicForm
                    phaseIdentifier={phaseIdentifier}
                    thematicId={thematicId}
                    editLocale={editLocale}
                    values={values}
                  />
                )}
                {section === '2' && <Export debateId={debateId} locale={locale} />}
              </AdminForm>
            </div>
            {steps.includes(section) && (
              <Navbar
                steps={steps}
                currentStep={section}
                totalSteps={3}
                phaseIdentifier={phaseIdentifier}
                beforeChangeSection={() => (pristine || submitting) && handleSubmit()}
              />
            )}
          </React.Fragment>
        );
      }}
    />
  );
};

const mapStateToProps = state => ({
  debateId: state.context.debateId,
  editLocale: state.admin.editLocale,
  locale: state.i18n.locale
});

export default compose(connect(mapStateToProps), withApollo)(DumbSurveyAdminForm);