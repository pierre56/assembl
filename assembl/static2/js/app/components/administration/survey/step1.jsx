// @flow
import React from 'react';
import { Field } from 'react-final-form';
import { I18n, Translate } from 'react-redux-i18n';
import { type ApolloClient, withApollo } from 'react-apollo';

import FieldArrayWithActions from '../../form/fieldArrayWithActions';
import MultilingualTextFieldAdapter from '../../form/multilingualTextFieldAdapter';
import { addThematicTooltip, deleteThematicTooltip, deleteSubThematicDisabledTooltip } from '../../common/tooltips';
import { removeMenuItem, addMenuItem, swapMenuItem } from '../thematicsMenu';

type Props = {
  discussionPhaseId: string,
  editLocale: string,
  locale: string,
  client: ApolloClient
};

const Step1 = ({ discussionPhaseId, editLocale, locale, client }: Props) => {
  const queryVariables = { discussionPhaseId: discussionPhaseId, lang: locale };
  return (
    <React.Fragment>
      <div className="form-title">{I18n.t('administration.survey.createTable')}</div>
      <FieldArrayWithActions
        isTree
        confirmDeletion
        name="themes"
        subFieldName="children"
        minItems={1}
        maxLevel={4}
        onRemove={id => removeMenuItem(id, client, queryVariables)}
        onAdd={(id, parentId, index) => addMenuItem(id, parentId, index, client, queryVariables)}
        onUp={(id, parentId, index, targetIndex) => swapMenuItem(id, parentId, index, targetIndex, client, queryVariables)}
        onDown={(id, parentId, index, targetIndex) => swapMenuItem(id, parentId, index, targetIndex, client, queryVariables)}
        renderFields={({ name, fieldIndex }) => (
          <Field
            required
            editLocale={editLocale}
            name={`${name}.title`}
            component={MultilingualTextFieldAdapter}
            label={`${I18n.t('administration.tableOfThematics.thematicTitle')} ${fieldIndex} ${editLocale.toUpperCase()}`}
          />
        )}
        tooltips={{
          addTooltip: addThematicTooltip,
          deleteTooltip: deleteThematicTooltip,
          deleteDisabled: deleteSubThematicDisabledTooltip
        }}
        confirmDeletionMessages={{
          confirmDeletionTitle: ({ index }) => (
            <Translate value="administration.tableOfThematics.confirmDeletionTitle" title={index} />
          ),
          confirmDeletionBody: ({ index }) => (
            <Translate value="administration.tableOfThematics.confirmDeletionBody" title={index} />
          )
        }}
      />
    </React.Fragment>
  );
};
export default withApollo(Step1);