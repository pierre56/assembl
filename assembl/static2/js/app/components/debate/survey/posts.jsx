/* eslint-disable  react/no-unused-prop-types */
// @flow
import * as React from 'react';
import { withRouter } from 'react-router';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import type { Map } from 'immutable';

import { updateContentLocale } from '../../../actions/contentLocaleActions';
import manageErrorAndLoading from '../../common/manageErrorAndLoading';
import FlatList from '../../common/flatList';
import Post from './post';
import QuestionPosts from '../../../graphql/QuestionPostsQuery.graphql';

type PostNode = {
  node: {
    id: string,
    originalLocale: string
  }
};

type Props = {
  isModerating: boolean,
  posts: {
    edges: Array<PostNode>
  },
  networkStatus: number,
  defaultContentLocaleMapping: Map,
  isPhaseCompleted: boolean,
  fetchMore: Function,
  refetch: Function,
  questionId: string,
  themeId: string,
  updateContentLocaleMapping: Function
};

export class DumbPosts extends React.Component<Props> {
  componentWillMount() {
    this.updateContentLocaleMappingFromProps(this.props);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.posts !== this.props.posts) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  updateContentLocaleMappingFromProps(props: Props) {
    const { defaultContentLocaleMapping, updateContentLocaleMapping, posts } = props;
    const contentLocaleMappingData = {};
    posts.edges.forEach((edge) => {
      const post = edge.node;
      const { id, originalLocale } = post;
      const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
      contentLocaleMappingData[id] = {
        contentLocale: contentLocale,
        originalLocale: originalLocale
      };
    });

    updateContentLocaleMapping(contentLocaleMappingData);
  }

  render() {
    const { isModerating, networkStatus, fetchMore, refetch, themeId, posts, questionId, isPhaseCompleted } = this.props;
    return (
      <FlatList
        items={posts}
        ListItem={Post}
        onEndReachedThreshold={0.6}
        extractItems={entities => entities.question.posts}
        networkStatus={networkStatus}
        fetchMore={fetchMore}
        refetch={refetch}
        loadPreviousMessage="debate.survey.loadRecentPosts"
        itemData={item => ({
          id: item.node.id,
          isPhaseCompleted: isPhaseCompleted,
          isModerating: isModerating,
          originalLocale: item.node.originalLocale,
          questionId: questionId,
          themeId: themeId
        })}
      />
    );
  }
}

const mapStateToProps = state => ({
  defaultContentLocaleMapping: state.defaultContentLocaleMapping,
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

const mapDispatchToProps = dispatch => ({
  updateContentLocaleMapping: data => dispatch(updateContentLocale(data))
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withRouter,
  graphql(QuestionPosts, {
    options: (props) => {
      const { hash } = props.location;
      let id = null;
      if (hash !== '') {
        id = hash.replace('#', '').split('?')[0];
      }
      return {
        variables: {
          first: 8,
          after: '',
          id: props.questionId,
          fromNode: id,
          isModerating: props.isModerating
        }
      };
    },
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading,
          networkStatus: data.networkStatus
        };
      }

      const { question: { posts }, fetchMore, refetch, networkStatus } = data;
      return {
        error: data.error,
        loading: data.loading,
        posts: posts,
        fetchMore: fetchMore,
        refetch: refetch,
        networkStatus: networkStatus
      };
    }
  }),
  manageErrorAndLoading({ color: 'black', displayLoader: true })
)(DumbPosts);