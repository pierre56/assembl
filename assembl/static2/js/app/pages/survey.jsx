// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Translate } from 'react-redux-i18n';
import { Grid } from 'react-bootstrap';
import type { Map } from 'immutable';

import { updateContentLocale } from '../actions/contentLocaleActions';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import Header from '../components/common/header';
import Question from '../components/debate/survey/question';
import Navigation from '../components/debate/survey/navigation';
import Proposals from '../components/debate/survey/proposals';
import { getIsPhaseCompletedById } from '../utils/timeline';
import ThematicQuery from '../graphql/ThematicQuery.graphql';
import { get as getRoute } from '../utils/routeMap';
import HeaderStatistics, { statContributions, statMessages, statParticipants } from '../components/common/headerStatistics';

type PostNode = {
  node: {
    id: string,
    originalLocale: string
  }
};

type QuestionType = {
  id: string,
  posts: {
    edges: Array<PostNode>
  },
  title: string
};

type Props = {
  timeline: Timeline,
  defaultContentLocaleMapping: Map, // eslint-disable-line react/no-unused-prop-types
  imgUrl: string,
  numContributors: number,
  numPosts: number,
  phaseId: string,
  questions: Array<QuestionType>,
  refetchThematic: Function,
  title: string,
  description: string,
  synthesisTitle: string,
  id: string,
  slug: string,
  totalSentiments: number,
  updateContentLocaleMapping: Function // eslint-disable-line react/no-unused-prop-types
};

type State = {
  isScroll: boolean,
  questionIndex: number | null,
  showModal: boolean
};

class Survey extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      isScroll: false,
      showModal: false,
      questionIndex: null
    };
  }

  componentWillMount() {
    this.updateContentLocaleMappingFromProps(this.props);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.questions !== this.props.questions) {
      this.updateContentLocaleMappingFromProps(nextProps);
    }
  }

  updateContentLocaleMappingFromProps(props) {
    const { defaultContentLocaleMapping, updateContentLocaleMapping, questions } = props;
    const contentLocaleMappingData = {};
    questions.forEach((question) => {
      question.posts.edges.forEach((edge) => {
        const post = edge.node;
        const { id, originalLocale } = post;
        const contentLocale = defaultContentLocaleMapping.get(originalLocale, originalLocale);
        contentLocaleMappingData[id] = {
          contentLocale: contentLocale,
          originalLocale: originalLocale
        };
      });
    });

    updateContentLocaleMapping(contentLocaleMappingData);
  }

  getIfProposals = (questions) => {
    if (!questions) return false;
    let isProposals = false;
    questions.forEach((question) => {
      if (question.posts.edges.length > 0) isProposals = true;
    });
    return isProposals;
  };

  scrollToQuestion = (isScroll, questionIndex) => {
    this.setState({
      isScroll: isScroll,
      questionIndex: questionIndex
    });
  };

  render() {
    const {
      id,
      imgUrl,
      numPosts,
      numContributors,
      phaseId,
      questions,
      refetchThematic,
      title,
      description,
      synthesisTitle,
      slug,
      totalSentiments,
      timeline
    } = this.props;

    const isPhaseCompleted = getIsPhaseCompletedById(timeline, phaseId);
    const phaseUrl = `${getRoute('debate', { slug: slug, phase: 'survey' })}`;
    let statElements = [];
    const numContributions = numPosts + totalSentiments;
    statElements = [statMessages(numPosts), statContributions(numContributions), statParticipants(numContributors)];
    return (
      <div className="survey">
        <div className="relative">
          <Header
            title={title}
            synthesisTitle={synthesisTitle}
            subtitle={description}
            imgUrl={imgUrl}
            phaseId={phaseId}
            type="idea"
          >
            <HeaderStatistics statElements={statElements} />
          </Header>
          <div className="questions">
            {questions &&
              questions.map((question, index) => (
                <Question
                  isPhaseCompleted={isPhaseCompleted}
                  title={question.title}
                  index={index + 1}
                  key={index}
                  questionsLength={questions.length}
                  questionId={question.id}
                  scrollToQuestion={this.scrollToQuestion}
                  refetchTheme={refetchThematic}
                />
              ))}
          </div>
          {questions && (
            <Navigation
              questionsLength={questions.length}
              questionIndex={this.state.questionIndex}
              isScroll={this.state.isScroll}
              scrollToQuestion={this.scrollToQuestion}
              isPhaseCompleted={isPhaseCompleted}
            />
          )}
          <div className="proposals" style={{ minHeight: '100px' }}>
            <section className={isPhaseCompleted ? 'shown' : 'proposals-section'} id="proposals">
              <Grid fluid className="background-light">
                {numPosts > 0 ? (
                  <div className="max-container">
                    <div className="question-title">
                      <div className="title-hyphen">&nbsp;</div>

                      <h1 className="dark-title-1">
                        <Translate value="debate.survey.proposalsTitle" />
                      </h1>
                    </div>

                    <div className="center">
                      {questions &&
                        questions.map((question, index) => (
                          <Proposals
                            nbPostsToShow={3}
                            questionsLength={questions.length}
                            themeId={id}
                            title={question.title}
                            posts={question.posts.edges}
                            questionIndex={index + 1}
                            questionId={question.id}
                            isPhaseCompleted={isPhaseCompleted}
                            phaseUrl={phaseUrl}
                            key={index}
                          />
                        ))}
                    </div>

                    <div className="margin-xl">&nbsp;</div>
                  </div>
                ) : null}
              </Grid>
            </section>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  timeline: state.timeline,
  defaultContentLocaleMapping: state.defaultContentLocaleMapping,
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

const mapDispatchToProps = dispatch => ({
  updateContentLocaleMapping: data => dispatch(updateContentLocale(data))
});

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(ThematicQuery, {
    props: ({ data, ownProps }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      const { thematic: { questions, numContributors, numPosts, totalSentiments }, refetch } = data;
      return {
        error: data.error,
        loading: data.loading,
        imgUrl: ownProps.headerImgUrl,
        numContributors: numContributors,
        numPosts: numPosts,
        questions: questions,
        refetchThematic: refetch,
        title: ownProps.title,
        description: ownProps.description,
        synthesisTitle: ownProps.synthesisTitle,
        totalSentiments: totalSentiments
      };
    }
  }),
  manageErrorAndLoading({ color: 'black', displayLoader: true })
)(Survey);