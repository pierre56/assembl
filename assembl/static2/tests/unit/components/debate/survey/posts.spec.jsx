import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { Map } from 'immutable';

import { DumbPosts } from '../../../../../js/app/components/debate/survey/posts';

describe('Posts component', () => {
  it('should display a flat list of posts', () => {
    const props = {
      updateContentLocaleMapping: () => null,
      defaultContentLocaleMapping: Map({ en: 'en' }),
      hasErrors: false,
      isModerating: false,
      posts: {
        edges: [
          {
            node: {
              title: 'Foobar'
            }
          }
        ]
      }
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbPosts {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});