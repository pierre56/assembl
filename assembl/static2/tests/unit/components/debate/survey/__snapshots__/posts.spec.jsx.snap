// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component should display a flat list of posts 1`] = `
<withRouter(DumbFlatList)
  ListItem={[Function]}
  extractItems={[Function]}
  itemData={[Function]}
  items={
    Object {
      "edges": Array [
        Object {
          "node": Object {
            "title": "Foobar",
          },
        },
      ],
    }
  }
  loadPreviousMessage="debate.survey.loadRecentPosts"
  onEndReachedThreshold={0.6}
/>
`;

exports[`Posts component should match snapshot 1`] = `
<withRouter(DumbFlatList)
  ListItem={[Function]}
  extractItems={[Function]}
  itemData={[Function]}
  items={
    Object {
      "edges": Array [
        Object {
          "node": Object {
            "title": "Foobar",
          },
        },
      ],
    }
  }
  loadPreviousMessage="debate.survey.loadRecentPosts"
  onEndReachedThreshold={0.6}
/>
`;
